File: leetcode/src/main/java/array/PascalsTriangle.java
package array;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/*-
    https://leetcode.com/problems/pascals-triangle/

    Cho số nguyên n, trả về n dòng của tam giác Pascal.

    Trong tam giác Pascal, mỗi số bằng tổng của hai số nằm trực tiếp phía trên nó,
    như hình minh họa:
 */
public class PascalsTriangle {

    public static List<List<Integer>> generate_ok(int numRows) {
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();
            // Phần tử đầu tiên luôn là 1
            row.add(1);
            // Các phần tử ở giữa
            for (int j = 1; j < i; j++) {
                List<Integer> integers = result.get(i - 1);
                int value = integers.get(j - 1) + integers.get(j);
                row.add(value);
            }
            // Phần tử cuối cùng (nếu không phải hàng đầu)
            if (i > 0) {
                row.add(1);
            }
            result.add(row);
        }
        return result;
    }

    public static List<List<Integer>> generate(int n) {
        List<List<Integer>> triangle = new ArrayList<>();
        // Tam giác Pascal có n hàng
        for (int i = 0; i < n; i++) {
            // Khởi tạo tất cả phần tử là 1
            List<Integer> row = new ArrayList<>(Collections.nCopies(i + 1, 1));
            Collections.fill(row, 1);
            // Cập nhật các phần tử ở giữa
            for (int j = 1; j < i; j++) {
                List<Integer> prevRow = triangle.get(i - 1);
                row.set(j, prevRow.get(j - 1) + prevRow.get(j));
            }
            triangle.add(row);
        }
        return triangle;
    }

    @Test
    public void test1() {
        // @formatter:off
        List<List<Integer>> result = List.of(
            List.of(1)
        );
        assertIterableEquals(result, generate(1));
        // @formatter:on
    }

    @Test
    public void test2() {
        // @formatter:off
        List<List<Integer>> result = List.of(
            List.of(1),
            List.of(1, 1),
            List.of(1, 2, 1),
            List.of(1, 3, 3, 1),
            List.of(1, 4, 6, 4, 1)
        );
        assertIterableEquals(result, generate(5));
        // @formatter:on
    }

}

File: py-algorithm/arrays/pascals_triangle.py
import unittest

from typing import List

"""
    https://leetcode.com/problems/pascals-triangle/
    
    Cho số nguyên n, trả về n dòng của tam giác Pascal.
    Trong tam giác Pascal, mỗi số bằng tổng của hai số nằm trực tiếp phía trên nó.
    
    Ý tưởng:
        - Hàng đầu tiên luôn là:
                [1]
        - Mỗi hàng bắt đầu và kết thúc bằng 1
            [1, ..., 1]
        - Các phần tử ở giữa được tính bằng:
            row[j] = prevRow[j−1] + prevRow[j]
"""


class Solution(unittest.TestCase):

    def generate(self, n: int) -> List[List[int]]:
        result = []
        # Tam giác có n dòng
        for i in range(n):
            # khởi tạo dòng toàn 1
            row = [1] * (i + 1)
            # Cập nhật các giá trị ở giữa (trừ 2 đầu mút)
            for j in range(1, i):
                previous_row = result[i - 1]
                row[j] = previous_row[j - 1] + previous_row[j]

            result.append(row)  # append sau khi hoàn thiện row

        return result

    @unittest.skip
    def test_1(self):
        self.assertEqual([[1]], self.generate(1))

    def test_2(self):
        self.assertEqual([[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]], self.generate(5))


if __name__ == '__main__':
    unittest.main()
